File: android/src/main/kotlin/com/therivanta/pixelcompressor/PixelCompressorPlugin.kt
package com.therivanta.pixelcompressor

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.therivanta.pixelcompressor.cache.CacheManager
import com.therivanta.pixelcompressor.capability.CapabilityProbe
import com.therivanta.pixelcompressor.errors.PixelCompressorErrors
import com.therivanta.pixelcompressor.errors.toPixelCompressorError
import com.therivanta.pixelcompressor.image.ImageEngine
import com.therivanta.pixelcompressor.metadata.MediaInfoEngine
import com.therivanta.pixelcompressor.progress.ProgressReporter
import com.therivanta.pixelcompressor.tasks.TaskRegistry
import com.therivanta.pixelcompressor.thumbnail.ThumbnailEngine
import com.therivanta.pixelcompressor.video.VideoEngine
import io.flutter.embedding.engine.plugins.FlutterPlugin
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

/** Registers pixel_compressor's platform channel HostApis and dispatches
 * each call to its engine class. This class stays dispatch only — no
 * compression logic lives here; it holds shared infra (task registry,
 * progress reporter, cache manager, capability probe) and every engine.
 *
 * Every HostApi method hands off to [pluginScope] (`Dispatchers.Default`)
 * immediately and returns without blocking the calling (platform) thread,
 * per the plugin's threading model. Every progress event and the final
 * Pigeon reply callback are hopped back to the main thread via
 * [mainHandler] before touching the generated Pigeon callers.
 */
class PixelCompressorPlugin :
  FlutterPlugin,
  ImageHostApi,
  VideoHostApi,
  ThumbnailHostApi,
  MetadataHostApi,
  CapabilityHostApi,
  CacheHostApi,
  TaskHostApi {

  private var pluginScope: CoroutineScope? = null
  private var taskRegistry: TaskRegistry? = null
  private var progressReporter: ProgressReporter? = null
  private var cacheManager: CacheManager? = null
  private var capabilityProbe: CapabilityProbe? = null
  private var imageEngine: ImageEngine? = null
  private var videoEngine: VideoEngine? = null
  private var thumbnailEngine: ThumbnailEngine? = null
  private var mediaInfoEngine: MediaInfoEngine? = null

  private val mainHandler = Handler(Looper.getMainLooper())

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    val context: Context = binding.applicationContext
    val messenger = binding.binaryMessenger

    val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    val registry = TaskRegistry()
    val reporter = ProgressReporter(PixelCompressorProgressCallbackApi(messenger, ""))
    val cache = CacheManager(context)
    val capability = CapabilityProbe()

    pluginScope = scope
    taskRegistry = registry
    progressReporter = reporter
    cacheManager = cache
    capabilityProbe = capability
    imageEngine = ImageEngine(context, cache, capability, reporter)
    videoEngine = VideoEngine(context, cache, capability, reporter)
    thumbnailEngine = ThumbnailEngine(context, cache, reporter)
    mediaInfoEngine = MediaInfoEngine()

    ImageHostApi.setUp(messenger, this)
    VideoHostApi.setUp(messenger, this)
    ThumbnailHostApi.setUp(messenger, this)
    MetadataHostApi.setUp(messenger, this)
    CapabilityHostApi.setUp(messenger, this)
    CacheHostApi.setUp(messenger, this)
    TaskHostApi.setUp(messenger, this)
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    val messenger = binding.binaryMessenger
    ImageHostApi.setUp(messenger, null)
    VideoHostApi.setUp(messenger, null)
    ThumbnailHostApi.setUp(messenger, null)
    MetadataHostApi.setUp(messenger, null)
    CapabilityHostApi.setUp(messenger, null)
    CacheHostApi.setUp(messenger, null)
    TaskHostApi.setUp(messenger, null)

    taskRegistry?.cancelAll()
    pluginScope?.cancel()
    pluginScope = null
    taskRegistry = null
    progressReporter = null
    cacheManager = null
    capabilityProbe = null
    imageEngine = null
    videoEngine = null
    thumbnailEngine = null
    mediaInfoEngine = null
  }

  // ---------------------------------------------------------------------
  // ImageHostApi / VideoHostApi / ThumbnailHostApi — registered, tracked,
  // progress-reporting tasks.
  // ---------------------------------------------------------------------

  override fun compressImage(
    request: ImageCompressRequest,
    callback: (Result<CompressionResultMessage>) -> Unit,
  ) {
    runTask(request.taskId, callback, terminalResult = { it }) {
      imageEngine!!.compress(request)
    }
  }

  override fun compressVideo(
    request: VideoCompressRequest,
    callback: (Result<CompressionResultMessage>) -> Unit,
  ) {
    runTask(request.taskId, callback, terminalResult = { it }) {
      videoEngine!!.compress(request)
    }
  }

  override fun generateThumbnails(
    request: ThumbnailRequest,
    callback: (Result<List<ThumbnailResultMessage>>) -> Unit,
  ) {
    runTask(request.taskId, callback, terminalResult = { null }) {
      thumbnailEngine!!.generate(request)
    }
  }

  // ---------------------------------------------------------------------
  // MetadataHostApi / CapabilityHostApi / CacheHostApi / TaskHostApi —
  // quick, non-cancellable, non-progress-reporting calls. Still
  // dispatched off the calling thread and replied to on the main thread,
  // per the shared threading model.
  // ---------------------------------------------------------------------

  override fun getMediaInfo(
    request: MediaInfoRequest,
    callback: (Result<MediaInfoMessage>) -> Unit,
  ) {
    runSimple(callback) { mediaInfoEngine!!.getMediaInfo(request) }
  }

  override fun capabilities(callback: (Result<CapabilitiesReportMessage>) -> Unit) {
    runSimple(callback) { capabilityProbe!!.capabilities() }
  }

  override fun clearCache(callback: (Result<Unit>) -> Unit) {
    runSimple(callback) { cacheManager!!.clear() }
  }

  override fun getCacheSize(callback: (Result<Long>) -> Unit) {
    runSimple(callback) { cacheManager!!.size() }
  }

  override fun cancel(taskId: String, callback: (Result<Boolean>) -> Unit) {
    runSimple(callback) { taskRegistry!!.cancel(taskId) }
  }

  override fun cancelAll(callback: (Result<Unit>) -> Unit) {
    runSimple(callback) { taskRegistry!!.cancelAll() }
  }

  override fun activeTaskIds(callback: (Result<List<String>>) -> Unit) {
    runSimple(callback) { taskRegistry!!.activeTaskIds() }
  }

  // ---------------------------------------------------------------------
  // Dispatch helpers
  // ---------------------------------------------------------------------

  /** Runs a registered, cancellable, progress-reporting task: registers
   * its [Job] in [TaskRegistry] under [taskId], emits exactly one
   * `onTaskTerminal` matching the outcome, and unregisters on completion.
   */
  private fun <T> runTask(
    taskId: String,
    callback: (Result<T>) -> Unit,
    terminalResult: (T) -> CompressionResultMessage?,
    block: suspend () -> T,
  ) {
    val scope = pluginScope
    val registry = taskRegistry
    val reporter = progressReporter
    if (scope == null || registry == null || reporter == null) {
      replyOnMain(callback, Result.failure(notAttachedError()))
      return
    }

    lateinit var job: Job
    job = scope.launch {
      try {
        val result = block()
        registry.unregister(taskId)
        reporter.onCompleted(taskId, terminalResult(result))
        replyOnMain(callback, Result.success(result))
      } catch (e: CancellationException) {
        registry.unregister(taskId)
        reporter.onCancelled(taskId)
        replyOnMain(callback, Result.failure(PixelCompressorErrors.cancelled()))
      } catch (e: Throwable) {
        registry.unregister(taskId)
        val mapped = e.toPixelCompressorError()
        reporter.onFailed(taskId, mapped.code, mapped.message)
        replyOnMain(callback, Result.failure(mapped))
      }
    }
    registry.register(taskId, job)
  }

  /** Runs a quick, non-cancellable call off the calling thread, replying
   * on the main thread. Used for metadata/capabilities/cache/task
   * queries that aren't tracked in [TaskRegistry] or reported through
   * [ProgressReporter].
   */
  private fun <T> runSimple(callback: (Result<T>) -> Unit, block: suspend () -> T) {
    val scope = pluginScope
    if (scope == null) {
      replyOnMain(callback, Result.failure(notAttachedError()))
      return
    }
    scope.launch {
      try {
        val result = block()
        replyOnMain(callback, Result.success(result))
      } catch (e: Throwable) {
        replyOnMain(callback, Result.failure(e.toPixelCompressorError()))
      }
    }
  }

  private fun <T> replyOnMain(callback: (Result<T>) -> Unit, result: Result<T>) {
    mainHandler.post { callback(result) }
  }

  private fun notAttachedError() =
    PixelCompressorErrors.platformNotSupported("PixelCompressorPlugin is not attached to a Flutter engine")
}

File: ios/pixel_compressor/Sources/pixel_compressor/PixelCompressorPlugin.swift
import Foundation

#if os(iOS)
  import Flutter
#elseif os(macOS)
  import FlutterMacOS
#else
  #error("Unsupported platform.")
#endif

/// Registers pixel_compressor's platform channel HostApis and dispatches
/// each call to the corresponding engine class. This class stays dispatch
/// only — no compression logic lives here (see `ImageEngine`,
/// `VideoEngine`, `ThumbnailEngine`, `MediaInfoEngine`, `CapabilityProbe`,
/// `CacheManager`).
///
/// Every HostApi method hands off to `workQueue` (concurrent, so an image
/// compress, a video compress, and a thumbnail batch can all run at once)
/// immediately and returns without blocking the calling thread, per the
/// Pigeon dispatch contract (no `TaskQueue` is configured, so the
/// generated dispatch itself runs on the main thread). Cancellable/
/// progress-emitting work (image/video/thumbnail) is wrapped in a
/// `DispatchWorkItem` registered with `TaskRegistry` so `TaskHostApi.cancel`
/// can cooperatively cancel it, and reports exactly one `onTaskTerminal`
/// event matching its Pigeon `completion` outcome. Every `ProgressReporter`
/// call and every Pigeon `completion` is hopped back to the main thread
/// before touching generated Pigeon callers.
public class PixelCompressorPlugin: NSObject, FlutterPlugin, ImageHostApi, VideoHostApi,
  ThumbnailHostApi, MetadataHostApi, CapabilityHostApi, TaskHostApi
{
  private let taskRegistry = TaskRegistry()
  private let cacheManager = CacheManager()
  private let progressReporter: ProgressReporter
  private let imageEngine: ImageEngine
  private let videoEngine: VideoEngine
  private let thumbnailEngine: ThumbnailEngine
  private let mediaInfoEngine = MediaInfoEngine()

  private let workQueue = DispatchQueue(
    label: "com.therivanta.pixel_compressor.engine", qos: .userInitiated, attributes: .concurrent)

  init(binaryMessenger: FlutterBinaryMessenger) {
    let callbackApi = PixelCompressorProgressCallbackApi(binaryMessenger: binaryMessenger)
    let reporter = ProgressReporter(api: callbackApi)
    progressReporter = reporter
    imageEngine = ImageEngine(
      cacheManager: cacheManager, taskRegistry: taskRegistry, progressReporter: reporter)
    videoEngine = VideoEngine(
      cacheManager: cacheManager, taskRegistry: taskRegistry, progressReporter: reporter)
    thumbnailEngine = ThumbnailEngine(
      cacheManager: cacheManager, taskRegistry: taskRegistry, progressReporter: reporter)
    super.init()
  }

  public static func register(with registrar: FlutterPluginRegistrar) {
    #if os(iOS)
      let messenger = registrar.messenger()
    #elseif os(macOS)
      let messenger = registrar.messenger
    #endif
    let instance = PixelCompressorPlugin(binaryMessenger: messenger)
    setUpAllApis(binaryMessenger: messenger, api: instance)
  }

  private static func setUpAllApis(binaryMessenger: FlutterBinaryMessenger, api: PixelCompressorPlugin) {
    ImageHostApiSetup.setUp(binaryMessenger: binaryMessenger, api: api)
    VideoHostApiSetup.setUp(binaryMessenger: binaryMessenger, api: api)
    ThumbnailHostApiSetup.setUp(binaryMessenger: binaryMessenger, api: api)
    MetadataHostApiSetup.setUp(binaryMessenger: binaryMessenger, api: api)
    CapabilityHostApiSetup.setUp(binaryMessenger: binaryMessenger, api: api)
    TaskHostApiSetup.setUp(binaryMessenger: binaryMessenger, api: api)
  }

  // MARK: - ImageHostApi / VideoHostApi / ThumbnailHostApi (cancellable, progress-emitting)

  func compressImage(
    request: ImageCompressRequest,
    completion: @escaping (Result<CompressionResultMessage, Error>) -> Void
  ) {
    dispatchCancellableTask(
      taskId: request.taskId,
      completion: completion,
      terminalResult: { $0 },
      work: { [imageEngine] in try imageEngine.compress(request: request) }
    )
  }

  func compressVideo(
    request: VideoCompressRequest,
    completion: @escaping (Result<CompressionResultMessage, Error>) -> Void
  ) {
    dispatchCancellableTask(
      taskId: request.taskId,
      completion: completion,
      terminalResult: { $0 },
      work: { [videoEngine] in try videoEngine.compress(request: request) }
    )
  }

  func generateThumbnails(
    request: ThumbnailRequest,
    completion: @escaping (Result<[ThumbnailResultMessage], Error>) -> Void
  ) {
    dispatchCancellableTask(
      taskId: request.taskId,
      completion: completion,
      terminalResult: { _ in nil },
      work: { [thumbnailEngine] in try thumbnailEngine.generate(request: request) }
    )
  }

  // MARK: - MetadataHostApi / CapabilityHostApi (fire-and-forget, no taskId)

  func getMediaInfo(
    request: MediaInfoRequest,
    completion: @escaping (Result<MediaInfoMessage, Error>) -> Void
  ) {
    workQueue.async { [mediaInfoEngine] in
      do {
        let result = try mediaInfoEngine.read(request: request)
        DispatchQueue.main.async { completion(.success(result)) }
      } catch {
        let mapped = PixelCompressorError.map(error)
        DispatchQueue.main.async { completion(.failure(mapped)) }
      }
    }
  }

  func capabilities(completion: @escaping (Result<CapabilitiesReportMessage, Error>) -> Void) {
    workQueue.async {
      let report = CapabilityProbe.probe()
      DispatchQueue.main.async { completion(.success(report)) }
    }
  }

  // MARK: - TaskHostApi

  func cancel(taskId: String, completion: @escaping (Result<Bool, Error>) -> Void) {
    completion(.success(taskRegistry.cancel(taskId: taskId)))
  }

  func cancelAll(completion: @escaping (Result<Void, Error>) -> Void) {
    taskRegistry.cancelAll()
    completion(.success(()))
  }

  func activeTaskIds(completion: @escaping (Result<[String], Error>) -> Void) {
    completion(.success(taskRegistry.activeTaskIds()))
  }

  // MARK: - Dispatch helper

  /// Wraps `work` in a `DispatchWorkItem`, registers it with `TaskRegistry`
  /// under `taskId` before dispatch (so a cancel racing the dispatch is
  /// never lost), runs it on `workQueue`, then unregisters and reports
  /// exactly one terminal event before hopping to the main thread to
  /// invoke `completion`. `terminalResult` extracts the
  /// `CompressionResultMessage` to attach to the terminal event from `T`
  /// (thumbnails have no single result message, so they pass `nil`).
  private func dispatchCancellableTask<T>(
    taskId: String,
    completion: @escaping (Result<T, Error>) -> Void,
    terminalResult: @escaping (T) -> CompressionResultMessage?,
    work: @escaping () throws -> T
  ) {
    let workItem = DispatchWorkItem { [weak self] in
      guard let self = self else { return }
      do {
        let result = try work()
        self.taskRegistry.unregister(taskId: taskId)
        self.progressReporter.reportTerminal(
          taskId: taskId, outcome: .completed, result: terminalResult(result))
        DispatchQueue.main.async { completion(.success(result)) }
      } catch {
        self.taskRegistry.unregister(taskId: taskId)
        let mapped = PixelCompressorError.map(error)
        let outcome: TaskOutcomeWire = mapped.code == "cancelled" ? .cancelled : .failed
        self.progressReporter.reportTerminal(
          taskId: taskId, outcome: outcome, errorCode: mapped.code, errorMessage: mapped.message)
        DispatchQueue.main.async { completion(.failure(mapped)) }
      }
    }
    taskRegistry.register(taskId: taskId, workItem: workItem)
    workQueue.async(execute: workItem)
  }
}
